src/data/access_control.rs

```diff
@@ -0,0 +1,217 @@
+    Insert,
+    /// Soft-update existing values.
+    Update,
+    /// Soft-delete existing values.
+    Delete,
+    /// Hard-update existing values.
+    HardUpdate,
+    /// Hard-delete existing values.
+    HardDelete,
+    /// Modify permissions for other users.
+    ModifyPermissions,
+}
+
+/// An AccessList consists of a list of users and their
+/// corresponding access configuration, as of a specific
+/// version of the data and the owners.
+/// The two flavours of data types have different
+/// variants of AccessList.
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub enum AccessList {
+    /// Denotes the public variant of an AccessList.
+    Public(PublicAccessList),
+    /// Denotes the private variant of an AccessList.
+    Private(PrivateAccessList),
+}
+
+impl From<PrivateAccessList> for AccessList {
+    fn from(list: PrivateAccessList) -> Self {
+        AccessList::Private(list)
+    }
+}
+
+impl From<PublicAccessList> for AccessList {
+    fn from(list: PublicAccessList) -> Self {
+        AccessList::Public(list)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub enum UserAccess {
+    Public(PublicUserAccess),
+    Private(PrivateUserAccess),
+}
+
+impl From<PrivateUserAccess> for UserAccess {
+    fn from(access: PrivateUserAccess) -> Self {
+        UserAccess::Private(access)
+    }
+}
+
+impl From<PublicUserAccess> for UserAccess {
+    fn from(access: PublicUserAccess) -> Self {
+        UserAccess::Public(access)
+    }
+}
+
+/// The access configuration to Private data, for a User.
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PrivateUserAccess {
+    config: BTreeMap<AccessType, bool>,
+}
+
+/// The access configuration to Public data, for a User.
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PublicUserAccess {
+    config: BTreeMap<AccessType, bool>,
+}
+
+impl PrivateUserAccess {
+    /// The ctor can be instantiated with
+    /// an access configuration.
+    pub fn new(config: BTreeMap<AccessType, bool>) -> Self {
+        PrivateUserAccess { config }
+    }
+
+    /// Determines if a specific access is
+    /// allowed according to this user access
+    /// configuration.
+    pub fn is_allowed(&self, access: AccessType) -> bool {
+        match self.config.get(&access) {
+            Some(true) => true,
+            _ => false,
+        }
+    }
+}
+
+impl PublicUserAccess {
+    /// The ctor can be instantiated with
+    /// an access configuration.
+    pub fn new(config: BTreeMap<AccessType, bool>) -> Self {
+        // todo: filter out Queries
+        PublicUserAccess { config }
+    }
+
+    /// Returns `Some(true)` if `access` is allowed and `Some(false)` if it's not.
+    /// `None` means that `User::Anyone` permissions apply.
+    pub fn is_allowed(&self, access: AccessType) -> Option<bool> {
```

Comment:
Well maybe so, even though I wouldn't say that is quite relevant or necessarily puts a limit on this usage. This is a ternary operation now. The `is_allowed` question is still valid because the answer can be **Yes**, **No** and **Maybe**.

See, the answer to a question is not always binary :) And that's what's modeled here.
Feel free to suggest something else, but I think this is neatly captured in the code.

Reply:
I would say it makes perfect sense though.
Are there rust standards saying that a question (that what "is_" denotes) cannot have answers Yes, No and Maybe?
It seems like a very strange an strict standard in that case. There must be some level of problem solving where the tools are used in ways where they solve problems, not only in the narrow scopes that someone before thought of.

This level seems more like a question of design and modelling, than a language standards area. But happy to hear about that standard, I haven't actually seen the list of those.

Reply:
Yea there doesn't seem to be a fixed set of standards as such. [I found this](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#predicates) which seems relevant and makes sense, i.e. variables / functions that represent / return `bool` are prefixed by a short question word.
To think of it a bit differently,
We can do things like `is_none()` on `Option`.
`app.get_permission().is_none()` seems to make more sense than `app.is_allowed().is_none()` IMO.
Happy to see what others think as well.

Reply:
Yeah, it says this
> Simple boolean predicates should be prefixed with is_ or another short question word, e.g., is_empty.

That's standard pretty much anywhere I would say.
But that practice _does not_ say that a more rich question, which can return the equivalent of a "maybe" answer, is not allowed or would be a bad practice. It simply does not regard that possibility, which IMO is just a lack of foreseeing. It's not a dictate or an opinion on it.
Coding stadards are full of this, and that's where the new standards are born, by those who fill in the gaps.

To the suggestion of yours, I would not consider it a good option because it is too technical. I'm very much convinced that good and clear code uses the naming to explain intention and domain meaning, not to describe the technical ways we solve it.
So to ask if it is "None", is simply not helpful.
The `.is_none()` method is not used on this return value, it is simply used in a match clause, where it is terse and simple.

We want to know _if it is allowed_, and the answer can be yes, no and maybe, in which case we dig a bit deeper to find the answer.

### To conclude
While there are probably other solutions as well, of which some could be better. I think this is _far too small of an issue_ to spend a lot of time on now. It's one of those things that could be added to a "code style doc" for whenever we have a joint session in the team to prioritize important code style questions and come to decisions.

Reply:
Agreed :+1: